File: finance_mobile/src/shared/api/whatIf.ts
import api, { buildPath } from "./client";
import type { WhatIfResult, WhatIfRequest } from "../types";

function path(userId: string) {
  return buildPath("/users/{user_id}/what-if", { user_id: userId });
}

export const whatIf = {
  simulate: (userId: string, period: string, data: WhatIfRequest) =>
    api
      .post<WhatIfResult>(path(userId), data, { params: { period } })
      .then((r) => r.data),
};

File: app/api/routers/what_if.py
from fastapi import APIRouter, HTTPException

from app.api.deps import DbSession, UserDep
from app.schemas.common import WhatIfRequest, WhatIfRead
from app.services import what_if_service
from app.services.what_if_service import NotFound

router = APIRouter(
    prefix="/users/{user_id}/what-if", tags=["what-if"]
)

_NOT_FOUND = {404: {"description": "Catégorie introuvable"}}


@router.post(
    "",
    response_model=WhatIfRead,
    responses=_NOT_FOUND,
)
async def what_if(period: str, payload: WhatIfRequest, user: UserDep, db: DbSession,):
    try:
        return await what_if_service.simulate(user.id, period, payload.category_id, payload.reduction_percent, db,)

    except NotFound as exc:
        raise HTTPException(
            status_code=404, detail=str(exc)
        ) from exc
